src/CoreWCF.Primitives/src/CoreWCF/Description/CustomEndpointAddressForMetadataBehavior.cs

```diff
@@ -0,0 +1,22 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.ObjectModel;
+using CoreWCF.Channels;
+
+namespace CoreWCF.Description
+{
+    public class CustomEndpointAddressForMetadataBehavior : IServiceBehavior
+    {
+        public IEndpointAddressForMetadataProvider Provider { get; }
+
+        public CustomEndpointAddressForMetadataBehavior(IEndpointAddressForMetadataProvider provider)
+        {
+            Provider = provider;
+        }
+
+        void IServiceBehavior.Validate(ServiceDescription serviceDescription, ServiceHostBase serviceHostBase) { }
+        void IServiceBehavior.AddBindingParameters(ServiceDescription serviceDescription, ServiceHostBase serviceHostBase, Collection<ServiceEndpoint> endpoints, BindingParameterCollection bindingParameters) { }
+        void IServiceBehavior.ApplyDispatchBehavior(ServiceDescription serviceDescription, ServiceHostBase serviceHostBase) { }
+    }
```

Comment:
Instead of simply being a container which is explicitly checked for by ServiceMetadataBehavior.ApplyDispatchBehavior, instead you could have this code directly apply the provider. Something like this:
```c#
var mexExt = host.Extensions.Find<ServiceMetadataExtension>();
mexExt.CustomEndpointAddressProvider = Provider;
```
To prevent problems with behavior order being a problem, you could use the internal api `ServiceMetadataExtension.EnsureServiceMetadataExtension(serviceHostBase)` to make sure an instance exists and return that instance instance of using host.Extensions.Find

Reply:
done